polyfill/RTCPeerConnection.js

```diff
@@ -1,56 +1,149 @@
+                            hostname: url.hostname,
+                            port: parseInt(url.port, 10),
+                            username: server.username,
+                            password: server.credential,
+                            // relayType - how to specify?
+                        };
+
+                        return iceServer;
+                    });
+                })
+                .flat(),
+            iceTransportPolicy: init?.iceTransportPolicy,
+        });
+
+        this.#peerConnection.onStateChange(() => {
+            this.dispatchEvent(new Event('connectionstatechange'));
+        });
+        this.#peerConnection.onSignalingStateChange(() => {
+            this.dispatchEvent(new Event('signalingstatechange'));
+        });
+        this.#peerConnection.onGatheringStateChange(() => {
+            this.dispatchEvent(new Event('icegatheringstatechange'));
+        });
+        this.#peerConnection.onDataChannel((channel) => {
+            this.dispatchEvent(new DataChannelEvent(new RTCDataChannel(channel)));
+        });
+
+        // forward events to properties
+        this.addEventListener('connectionstatechange', this.onconnectionstatechange);
+        this.addEventListener('signalingstatechange', this.onsignalingstatechange);
+        this.addEventListener('icegatheringstatechange', this.onicegatheringstatechange);
+        this.addEventListener('datachannel', this.ondatachannel);
+
+        this.#peerConnection.onLocalDescription((sdp, type) => {
+            if (type === 'offer') {
+                this.#localOffer.resolve({ sdp, type });
+            }
+
+            if (type === 'answer') {
+                this.#localAnswer.resolve({ sdp, type });
+            }
+        });
+
+        this.#peerConnection.onLocalCandidate((candidate, mid) => {
+            if (mid === 'unspec') {
+                this.#localAnswer.reject(new Error(`Invalid description type ${mid}`));
+                return;
+            }
+
+            const event = new PeerConnectionIceEvent(new RTCIceCandidate({ candidate }));
+
+            this.onicecandidate(event);
+        });
+    }
 
     #canTrickleIceCandidates = null;
     get canTrickleIceCandidates() {
         return this.#canTrickleIceCandidates;
     }
-    #connectionState = null;
+
     get connectionState() {
-        return this.#connectionState;
+        return this.#peerConnection.state();
     }
-    #currentLocalDescription = null;
+
     get currentLocalDescription() {
-        return this.#currentLocalDescription;
+        return this.#peerConnection.localDescription();
     }
-    #currentRemoteDescription = null;
+
     get currentRemoteDescription() {
-        return this.#currentRemoteDescription;
+        // not exposed by node-datachannel
+        return toSessionDescription(null);
     }
-    #iceConnectionState = null;
+
     get iceConnectionState() {
-        return this.#iceConnectionState;
+        return this.#peerConnection.state();
```

Comment:
This returns the same underlying prop for `.connectionState` and `iceConnectionState` - I wasn't sure which one to use on the underlying `PeerConnection` but we don't use this in `@libp2p/webrtc` so it wasn't a big deal.

It would need fixing here though.

Reply:
yup, the type checks error on this, so it's visible to us that it needs to be changed

Reply:
I can add `iceConnectionState` to libdatachannel, is there something else missing?

Reply:
[ice restart](https://github.com/murat-dogan/node-datachannel/blob/00ed874ad67b95099ca56baf0f9e86fd5c5d337d/polyfill/RTCPeerConnection.js#L219), [Receivers](https://github.com/murat-dogan/node-datachannel/blob/00ed874ad67b95099ca56baf0f9e86fd5c5d337d/polyfill/RTCPeerConnection.js#L204C8-L204C17), [Senders](https://github.com/murat-dogan/node-datachannel/blob/00ed874ad67b95099ca56baf0f9e86fd5c5d337d/polyfill/RTCPeerConnection.js#L207C8-L207C8), [Transceivers](https://github.com/murat-dogan/node-datachannel/blob/00ed874ad67b95099ca56baf0f9e86fd5c5d337d/polyfill/RTCPeerConnection.js#L213C8-L213C20), and maybe stats?

Reply:
About Receivers, Senders, and Transceivers, do you plan to polyfill tracks? I don't think this is possible currently, given tracks are lower level in libdatachannel and basically transceivers themselves (libdatachannel doesn't handle media encoding/decoding).

Reply:
I don't know? I haven't even had time to test this as libdatachannel segfaults on my os and I haven't had time to debug it, we'd need to implement stuff from https://devdocs.io/dom/rtcstats/type
and I don't know what we'd need to track ourselves and what exists in libdatachannel

Reply:
polyfilling tracks means polyfilling mediasource, so pretty much faking it like mediasource.fromFilePath or some other non standard stuff, and a transceiver needs to exist to add a track, so then that might need to be faked too, idk dont have the time to check

this pr is just for the bare minimum simple functionality, so for now it's not something we need to worry about, but generating certificates and getting ICE data would be nice

Reply:
FYI I added the ICE connection state to libdatachannel in https://github.com/paullouisageneau/libdatachannel/pull/944